blockchain/src/codec.rs

```diff
@@ -0,0 +1,388 @@
+    }
+
+    fn encoded_length(&self) -> usize {
+        8 + self.tip.encoded_length() + 32 + 8 + 4 + self.shortid_list.len()
+    }
+}
+
+impl Decodable for Inventory {
+    fn decode(buf: &mut impl Reader) -> Result<Self, ReadError> {
+        Ok(Inventory {
+            version: buf.read_u64()?,
+            tip: BlockHeader::decode(buf)?,
+            tip_signature: buf.read_signature()?,
+            shortid_nonce: buf.read_u64()?,
+            shortid_list: buf.read_shortid_vec()?,
+        })
+    }
+}
+
+// Tx now implement Encoder, but it's implementation isn't suitable for
+// encoding in that case
+fn read_tx(buf: &mut impl Reader) -> Result<Tx, ReadError> {
+    let tx_size = buf.read_u32()? as usize;
+    let vec = buf.read_vec(tx_size)?;
+    Tx::from_bytes(&vec).map_err(|_| ReadError::InvalidFormat)
+}
+
+fn write_tx(tx: &Tx, dst: &mut impl Writer) -> Result<(), WriteError> {
+    dst.write_u32(b"tx length", tx.encoded_length() as u32)?;
+    tx.encode(dst)
+}
+
+fn read_block_tx(src: &mut impl Reader) -> Result<BlockTx, ReadError> {
+    let tx = read_tx(src)?;
+    let len = src.read_u32()? as usize;
+    let proofs = src.read_vec_with(len, 1, |src| match src.read_u8()? {
+        0 => Ok(Proof::Transient),
+        1 => merkle::Path::decode(src)
+            .map(Proof::Committed)
+            .map_err(|_| ReadError::InvalidFormat),
+        _ => Err(ReadError::InvalidFormat),
+    })?;
+    Ok(BlockTx { tx, proofs })
+}
+
+fn write_block_tx(block: &BlockTx, dst: &mut impl Writer) -> Result<(), WriteError> {
+    write_tx(&block.tx, dst)?;
+    dst.write_u32(b"n", block.proofs.len() as u32)?;
+    for proof in block.proofs.iter() {
+        match proof {
+            Proof::Transient => dst.write_u8(b"type", 0)?,
+            Proof::Committed(path) => {
+                dst.write_u8(b"type", 1)?;
+                path.encode(dst)?;
+            }
+        }
+    }
+    Ok(())
+}
+
+fn read_block_txs(src: &mut impl Reader) -> Result<Vec<BlockTx>, ReadError> {
+    let len = src.read_u32()? as usize;
+    const BLOCK_TX_MIN_LENGTH: usize = 8 + 8 + 8 + 8 + 32 + 32 + 1 + 11 * 32 + 8;
+    src.read_vec_with(len, BLOCK_TX_MIN_LENGTH, read_block_tx)
+}
+
+fn write_block_txs(block_txs: &[BlockTx], dst: &mut impl Writer) -> Result<(), WriteError> {
+    dst.write_u32(b"block_txs length", block_txs.len() as u32)?;
+    block_txs.iter().map(|e| write_block_tx(e, dst)).collect()
+}
+
+trait ReaderExt: Reader + Sized {
+    fn read_u8_vec(&mut self) -> Result<Vec<u8>, ReadError> {
+        let len = self.read_u32()? as usize;
+        self.read_vec(len)
+    }
+
+    fn read_signature(&mut self) -> Result<Signature, ReadError> {
+        let s = self.read_u8x32()?;
+        let r = self.read_u8x32()?;
+        Ok(Signature {
+            s: Scalar::from_bits(s),
+            R: CompressedRistretto(r),
+        })
+    }
+
+    fn read_shortid_vec(&mut self) -> Result<ShortIDVec, ReadError> {
+        self.read_u8_vec().map(ShortIDVec)
```

Comment:
I would make vec private so we don't have trailing bytes and mess up encoding length calculation, and make a fallible `new(Vec<u8>)->Option<ShortIDVec>` method:

```suggestion
        self.read_u8_vec().map(ShortIDVec::new).ok_or(ReadError::InsufficientBytes).
```
```suggestion
        self.read_u8_vec().map(ShortIDVec)
```
```suggestion
        self.read_u8_vec().map(ShortIDVec)
```

(we can do it in a separate PR)

Reply:
Cf. https://github.com/stellar/slingshot/issues/436